### PR TITLE
Implement archive malware scan for local manifests

### PR DESCRIPTION
**Pull request by ryfu-msft**
Changes:

- Added project files to build the PureLib library and integrate with our current workflow to scan archive files downloaded from local manifests. 
- Updates NOTICE.txt

Tests:
- Added workflow tests using testhooks to verify that the correct behavior occurs if scanning fails.
- Added a unit test to check whether a valid zip is scanned successfully. 
- Manually verified that scanning can detect malware using a bad zip file.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2566)

### CODE REVIEW

File: src/AppInstallerCLI.sln [hunk 1]
```diff
@@ -1026,6 +1028,46 @@ Global
 		{787EC629-C0FB-4BA9-9746-4A82CD06B73E}.TestRelease|x64.Build.0 = Release|x64
 		{787EC629-C0FB-4BA9-9746-4A82CD06B73E}.TestRelease|x86.ActiveCfg = Debug|Win32
 		{787EC629-C0FB-4BA9-9746-4A82CD06B73E}.TestRelease|x86.Build.0 = Debug|Win32
+		{BB14D603-F44E-4415-8770-BF3E13F4C17F}.Debug|Any CPU.ActiveCfg = Debug|x64
+		{BB14D603-F44E-4415-8770-BF3E13F4C17F}.Debug|Any CPU.Build.0 = Debug|x64
+		{BB14D603-F44E-4415-8770-BF3E13F4C17F}.Debug|ARM.ActiveCfg = Debug|x64
+		{BB14D603-F44E-4415-8770-BF3E13F4C17F}.Debug|ARM.Build.0 = Debug|x64
```
Comment (yao-msft) [inline src/AppInstallerCLI.sln:1034]: All these configurations probably need to be revisited

File: src/WindowsPackageManager/WindowsPackageManager.vcxproj [hunk 2]
```diff
@@ -317,6 +317,9 @@
     <ProjectReference Include="..\YamlCppLib\YamlCppLib.vcxproj">
       <Project>{8bb94bb8-374f-4294-bca1-c7811514a6b7}</Project>
     </ProjectReference>
+    <ProjectReference Include="..\PureLib\PureLib.vcxproj">
+      <Project>{8bb94bb8-374f-4294-bca1-c7811514a6b7}</Project>
```
Comment (yao-msft) [inline src/WindowsPackageManager/WindowsPackageManager.vcxproj:321]: >{8bb94bb8-374f-4294-bca1-c7811514a6b7} [](http://example.com/codeflow?start=15&length=38)

probably wrong value, same as yamlcpplib above

File: src/AppInstallerCLITests/AppInstallerCLITests.vcxproj [hunk 3]
```diff
@@ -796,6 +796,9 @@
     <ProjectReference Include="..\YamlCppLib\YamlCppLib.vcxproj">
       <Project>{8bb94bb8-374f-4294-bca1-c7811514a6b7}</Project>
     </ProjectReference>
+    <ProjectReference Include="..\PureLib\PureLib.vcxproj">
+      <Project>{af2f39c5-3c3a-4344-bd38-3ac32478dcaa}</Project>
```
Comment (yao-msft) [inline src/AppInstallerCLITests/AppInstallerCLITests.vcxproj:800]: >{af2f39c5-3c3a-4344-bd38-3ac32478dcaa} [](http://example.com/codeflow?start=15&length=38)

{BB14D603-F44E-4415-8770-BF3E13F4C17F}

File: src/AppInstallerCommonCore/Archive.cpp [hunk 4]
```diff
@@ -1,6 +1,13 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 #include "pch.h"
+#include "winget/Archive.h"
+#pragma warning( push )
+#pragma warning ( disable : 4189 )
+#pragma warning ( disable : 4244 )
+#pragma warning ( disable : 26451 )
```
Comment (yao-msft) [inline src/AppInstallerCommonCore/Archive.cpp:8]: nit: disable them all in 1 line

File: src/AppInstallerCLICore/Workflows/ArchiveFlow.cpp [hunk 5]
```diff
@@ -15,6 +15,34 @@ namespace AppInstaller::CLI::Workflow
...
+            {
+                // TODO: replace with proper --force argument when available.
+                if (context.Args.Contains(Execution::Args::Type::HashOverride))
+                {
+                    AICLI_LOG(CLI, Info, << "Archive malware scan failed; proceeding due to --force override");
+                    context.Reporter.Error() << Resource::String::ArchiveFailedMalwareScanOverridden << std::endl;
+                }
+                else
+                {
+                    AICLI_LOG(CLI, Info, << "Archive malware scan failed");
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/ArchiveFlow.cpp:38]: >Info [](http://example.com/codeflow?start=35&length=4)

Error

File: src/AppInstallerCLICore/Workflows/ArchiveFlow.cpp [hunk 6]
```diff
@@ -15,6 +15,34 @@ namespace AppInstaller::CLI::Workflow
...
+            if (scanResult)
+            {
+                AICLI_LOG(CLI, Info, << "Archive malware scan passed");
+            }
+            else
+            {
+                // TODO: replace with proper --force argument when available.
+                if (context.Args.Contains(Execution::Args::Type::HashOverride))
+                {
+                    AICLI_LOG(CLI, Info, << "Archive malware scan failed; proceeding due to --force override");
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/ArchiveFlow.cpp:33]: >Info [](http://example.com/codeflow?start=35&length=4)

Warning?

File: src/AppInstallerCLICore/Workflows/ArchiveFlow.cpp [hunk 7]
```diff
@@ -15,6 +15,34 @@ namespace AppInstaller::CLI::Workflow
...
+                    context.Reporter.Error() << Resource::String::ArchiveFailedMalwareScanOverridden << std::endl;
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/ArchiveFlow.cpp:34]: >Error() [](http://example.com/codeflow?start=37&length=7)

Warning if it's overriden?

File: src/AppInstallerCommonCore/Archive.cpp [hunk 8]
```diff
@@ -1,6 +1,13 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 #include "pch.h"
+#include "winget/Archive.h"
+#pragma warning( push )
+#pragma warning ( disable : 4189 )
+#pragma warning ( disable : 4244 )
+#pragma warning ( disable : 26451 )
+#include "pure.h"
```
Comment (yao-msft) [inline src/AppInstallerCommonCore/Archive.cpp:9]: pure.h should probably go to pch.h

File: src/AppInstallerCommonCore/Archive.cpp [hunk 9]
```diff
@@ -41,4 +48,33 @@ namespace AppInstaller::Archive
...
+    {
+#ifndef AICLI_DISABLE_TEST_HOOKS
+        if (s_ScanArchiveResult_TestHook_Override)
+        {
+            return *s_ScanArchiveResult_TestHook_Override;
+        }
+#endif
+
+        std::ifstream instream(zipPath, std::ios::in | std::ios::binary);
+        std::vector<uint8_t> data((std::istreambuf_iterator<char>(instream)), std::istreambuf_iterator<char>());
```
Comment (yao-msft) [inline src/AppInstallerCommonCore/Archive.cpp:72]: >std::vector<uint8_t> data((std::istreambuf_iterator<char>(instream)), std::istreambuf_iterator<char>()); [](http://example.com/codeflow?start=8&length=104)

nit:
std::vector<uint8_t> data{ std::istreambuf_iterator<char>(instream), std::istreambuf_iterator<char>() };

File: src/PureLib/PureLib.vcxproj [hunk 10]
```diff
@@ -0,0 +1,225 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup Label="Globals">
+    <VCProjectVersion>16.0</VCProjectVersion>
```
Comment (yao-msft) [inline src/PureLib/PureLib.vcxproj:4]: >16.0 [](http://example.com/codeflow?start=22&length=4)

nit: maybe set to 15.0 to be consistent with other existing projects

File: src/PureLib/PureLib.vcxproj [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+      <Platform>ARM64</Platform>
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Debug|Win32">
+      <Configuration>Debug</Configuration>
+      <Platform>Win32</Platform>
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Fuzzing|x64">
+      <Configuration>Fuzzing</Configuration>
+      <Platform>x64</Platform>
+    </ProjectConfiguration>
```
Comment (yao-msft) [inline src/PureLib/PureLib.vcxproj:28]: remove, fuzzing configuration is not needed for this project

File: src/PureLib/PureLib.vcxproj [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
+    <LinkIncremental>true</LinkIncremental>
+    <OutDir>$(SolutionDir)x86\$(Configuration)\$(ProjectName)\</OutDir>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
+    <OutDir>$(SolutionDir)$(Platform)\$(Configuration)\$(ProjectName)\</OutDir>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Fuzzing|x64'">
+    <OutDir>$(SolutionDir)$(Platform)\$(Configuration)\$(ProjectName)\</OutDir>
+  </PropertyGroup>
```
Comment (yao-msft) [inline src/PureLib/PureLib.vcxproj:77]: remove

File: src/PureLib/PureLib.vcxproj [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+  </ItemDefinitionGroup>
+  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Fuzzing|x64'">
+    <ClCompile>
+      <LanguageStandard>stdcpp17</LanguageStandard>
+      <AdditionalIncludeDirectories>pure</AdditionalIncludeDirectories>
+      <WarningLevel>Level3</WarningLevel>
+      <AdditionalOptions>/fsanitize-coverage=inline-8bit-counters /fsanitize-coverage=edge /fsanitize-coverage=trace-cmp /fsanitize-coverage=trace-div %(AdditionalOptions)</AdditionalOptions>
+      <RuntimeLibrary>MultiThreaded</RuntimeLibrary>
+    </ClCompile>
+  </ItemDefinitionGroup>
```
Comment (yao-msft) [inline src/PureLib/PureLib.vcxproj:141]: remove

File: src/PureLib/PureLib.vcxproj [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+      <AdditionalIncludeDirectories>pure;pure\zlib;</AdditionalIncludeDirectories>
+      <WarningLevel>Level3</WarningLevel>
+    </ClCompile>
+    <ProjectReference>
+      <LinkLibraryDependencies>false</LinkLibraryDependencies>
+    </ProjectReference>
+    <PreBuildEvent>
+      <Command>
+      </Command>
+    </PreBuildEvent>
```
Comment (yao-msft) [inline src/PureLib/PureLib.vcxproj:182]: seems like not needed? Only exists in debug|x64 configuration

File: src/AppInstallerCommonCore/Archive.cpp (hunk 8 above)
Comment (ryfu-msft) [reply src/AppInstallerCommonCore/Archive.cpp]: Discussed offline, leaving this as is. Added a todo statement to move this to pch.h in the future.

File: src/PureLib/PureLib.vcxproj [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+    <ClCompile>
+      <LanguageStandard>stdcpp17</LanguageStandard>
+      <AdditionalIncludeDirectories>pure;pure\zlib;</AdditionalIncludeDirectories>
+      <WarningLevel>Level3</WarningLevel>
+      <BasicRuntimeChecks>Default</BasicRuntimeChecks>
+      <RuntimeLibrary>MultiThreadedDebugDLL</RuntimeLibrary>
+    </ClCompile>
+    <ProjectReference>
+      <LinkLibraryDependencies>false</LinkLibraryDependencies>
+    </ProjectReference>
```
Comment (yao-msft) [inline src/PureLib/PureLib.vcxproj:164]: Is this part needed?

**Pull request merged**